letsencrypt/auth_handler.py

```diff
@@ -113,6 +113,16 @@ def _solve_challenges(self):
                     cont_resp = self.cont_auth.perform(self.cont_c)
                 if self.dv_c:
                     dv_resp = self.dv_auth.perform(self.dv_c)
+                    for response in dv_resp:
+                        if isinstance(response, challenges.HTTP01Response):
+                            for achall in self.dv_c:
+                                if(isinstance(achall.challb.chall, challenges.HTTP01) or
+                                    isinstance(achall.challb.chall, challenges.TLSSNI01)):
```

Comment:
That's wrong logic, because:
- it tries to simple verify (almost) all possible pairs of challenges and responses
- tries to match `http-01` response with `tls-sni-01` challenge

You should rather use `for chall, resp in zip(challenges, responses)`. `interfaces.py` documentation should be updated to make it explicit that plugins are required to return responses in the same order as challenges.